zulip_bots/zulip_bots/test_lib.py

```diff
@@ -90,21 +91,43 @@ def check_expected_responses(self, expectations, expected_method='send_reply',
                 # For calls with send_message, r is already a Dict.
                 message = dict(message_template, content = m)
                 response = {'content': r} if expected_method == 'send_reply' else r
-                self.assert_bot_response(message=message, response=response,
-                                         expected_method=expected_method)
+                self.assert_bot_responses(message, (response, expected_method))
 
-    def call_request(self, message, expected_method, response):
-        # type: (Dict[str, Any], str, Dict[str, Any]) -> None
-        # Send message to the concerned bot
-        self.message_handler.handle_message(message, self.mock_bot_handler)
+    def call_request(self, message, *responses):
+        # type: (Dict[str, Any], *Tuple[Dict[str, Any], str]) -> None
 
-        # Check if the bot is sending a message via `send_message` function.
-        # Where response is a dictionary here.
+        # Mock BotHandler; get instance
+        bot_handler = self.MockClass()
         instance = self.MockClass.return_value
-        if expected_method == "send_message":
-            instance.send_message.assert_called_with(response)
-        else:
-            instance.send_reply.assert_called_with(message, response['content'])
+
+        # Send message to the bot
+        try:
+            self.message_handler.handle_message(message, self.mock_bot_handler)
+        except KeyError as key_err:
+            raise Exception("Message tested likely required key {}.".format(key_err))
+
+        # Determine which messaging functions are expected
+        send_messages = [call(r[0]) for r in responses if r[1] == 'send_message']
+        send_replies = [call(message, r[0]['content']) for r in responses if r[1] == 'send_reply']
+
+        # Test that call were matching in quantity, and then in details
+        fail_template = "\nMESSAGE:\n{}\nACTUAL CALLS:\n{}\nEXPECTED:\n{}\n"
+        functions_to_test = (('send_message', instance.send_message, send_messages),
+                             ('send_reply', instance.send_reply, send_replies))
+        for version, actual, expected in functions_to_test:
+            assert len(expected) == actual.call_count, (
+                "Numbers of '{}' called do not match those expected ({} calls, {} expected)" +
+                fail_template).format(version, actual.call_count, len(expected),
+                                      message, actual.call_args_list, expected)
+            try:
+                if len(expected) > 0:
+                    actual.assert_has_calls(expected)  # any_order = True FIXME
```

Comment:
We should probably just pass `any_order=True`, right?

Reply:
I *think* that should be right, but I wasn't sure when working on it, though I believe so now.

I'm wondering if this would be cleaner and more concise as:
```python
if len(expected) > 0:
    if set(expected) != set(actual.call_args_list):
        raise AssertionError("...")
    actual.reset_mock()
```
That avoids the slightly strange re-raise and is more explicit?